src/explore-education-statistics-frontend/server.js

```diff
@@ -84,32 +84,62 @@ async function startServer() {
     return `${input.slice(0, -pattern.length)}${replacement}`;
   }
 
+  function isRedirectionRequired(request) {
+    const hostnameIncludesWWW = request.hostname.startsWith('www');
+    let redirectionPath = request.path;
+
+    // Redirect URLs with trailing slash to equivalent without slash with 301
+    redirectionPath = replaceLastOccurrence(redirectionPath, '/', '');
+
+    // Bots are adding /1000 and Google isn't yet omitting them from its index automatically,
+    // so we should redirect away from them
+    redirectionPath = replaceLastOccurrence(redirectionPath, '/1000', '');
+
+    // Search wrongly indexed pages from Google Search Console for matches
+    // Redirect away if found to (eventually) clear routes from index
+    const urlMatch = seoRedirects.find(source => source.from === request.url);
+    if (urlMatch !== undefined) {
+      redirectionPath.path = urlMatch.to;
+    }
+
+    const redirectionRequired =
+      hostnameIncludesWWW || redirectionPath !== request.path;
+
+    if (redirectionRequired) {
+      // PUBLIC_URL ends in a slash, and redirectionPath starts with one, but we only want one in total
+      const newUrl = new URL(
+        `${replaceLastOccurrence(
+          process.env.PUBLIC_URL,
+          '/',
+          '',
+        )}${redirectionPath}`,
+      );
+
+      // Restore any search parameters on original request
+      Object.keys(request.query).forEach(paramKey => {
+        newUrl.searchParams.append(paramKey, request.param(paramKey));
+      });
+
+      return newUrl.href;
+    }
+
+    return undefined;
+  }
+
   server.use((req, res, nextFunc) => {
+    // Return early for speed if redirect definitely isn't required
     if (
-      req.url === '/' ||
+      (req.url === '/' && !req.hostname.startsWith('www')) ||
       req.url.startsWith('/assets') ||
       req.url.startsWith('/_next')
     ) {
       nextFunc();
       return undefined;
     }
 
-    let newUri = req.url;
-    // Redirect URLs with trailing slash to equivalent without slash with 301
-    newUri = replaceLastOccurrence(newUri, '/', '');
-
-    // These are generated by spam sites. We could redirect them too to reduce Google's indexed 404 pages,
-    // but it seems current advice is actually to ignore these:
-    // https://support.google.com/webmasters/thread/253569816/google-crawler-adding-a-1000-to-the-end-of-a-ton-of-urls?hl=en
-    // newUri = replaceLastOccurance(newUri, '/1000', '');
-
-    const urlMatch = seoRedirects.find(source => source.from === req.url);
-    if (urlMatch !== undefined) {
-      newUri = urlMatch.to;
-    }
-
-    if (newUri !== req.url) {
-      return res.redirect(301, newUri);
+    const potentialRedirectUrl = isRedirectionRequired(req);
+    if (potentialRedirectUrl !== undefined) {
```

Comment:
Similar to other comments, could simplify to `if (redirectUrl)`

Reply:
Renamed